### PR TITLE
cmake: Use `pkg-config --static` when building with depends

### PR DESCRIPTION
**Pull request by hebasto**
This PR mirrors the current master branch behaviour defined in https://github.com/hebasto/bitcoin/blob/53f4607cc8c67366662f49cb312d2e4ff8b6523a/depends/config.site.in#L85

Also the `cross_pkg_check_modules` wrapper around the standard `pkg_check_modules` command has been removed.

This change became possible after bumping CMake minimum required version up to 3.22.

### CODE REVIEW

**Comment by hebasto** [comment]
It was pointed by @theuni in the past. However, I can't remember the exact circumstances :)

File: cmake/module/AddLibeventIfNeeded.cmake [hunk 1]
```diff
@@ -46,12 +46,11 @@ function(add_libevent_if_needed)
...
     libevent>=${libevent_minimum_version}
   )
-  if(MINGW)
-    target_link_libraries(PkgConfig::libevent INTERFACE
-      iphlpapi
-      ws2_32
-    )
-  endif()
+  target_link_libraries(PkgConfig::libevent INTERFACE
+    # Due to a bug in the libevent build system, the required
```
Comment (fanquake) [inline cmake/module/AddLibeventIfNeeded.cmake:50]: Is this tracked upstream somewhere, have we submitted a fix?

**Review by fanquake — COMMENTED** [review]
Can you add a PR description? What does this fix or improve? How does this change the binaries from what we are already producing?

**Comment by theuni** [comment]
`pkg-config --static` tells pkg-config to include Libs.private entries as well. For shared libs, these would be linked in already, but for shared libs they need to be included at the final link stage.

This isn't necessarily needed for cross builds, but it's needed for depends because all libs are built statically there. It's currently added [here](https://github.com/bitcoin/bitcoin/blob/master/depends/config.site.in#L85).

Can we not hack this into the generated toolchain file?

**Comment by hebasto** [comment]
@theuni 

> This isn't necessarily needed for cross builds, but it's needed for depends because all libs are built statically there. It's currently added [here](https://github.com/bitcoin/bitcoin/blob/master/depends/config.site.in#L85).
> 
> Can we not hack this into the generated toolchain file?

Thank you! Your suggestions have been implemented.

File: cmake/module/AddLibeventIfNeeded.cmake (hunk 1 above)
Comment (hebasto) [reply cmake/module/AddLibeventIfNeeded.cmake:50]: I've added a descriptive comment.

Comment (fanquake) [reply cmake/module/AddLibeventIfNeeded.cmake:50]: Shouldn't this be getting patched into depends?

**Comment by hebasto** [comment]
All comments have been addressed per feedback.

File: depends/toolchain.cmake.in [hunk 2]
```diff
@@ -111,6 +109,15 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Darwin" AND NOT CMAKE_HOST_APPLE)
   set(CMAKE_FRAMEWORK_PATH "@OSX_SDK@/System/Library/Frameworks")
 endif()
 
+
+# Customize pkg-config behaviour.
+cmake_path(APPEND CMAKE_FIND_ROOT_PATH "lib" "pkgconfig" OUTPUT_VARIABLE pkg_config_path)
+set(ENV{PKG_CONFIG_PATH} ${pkg_config_path})
```
Comment (theuni) [inline depends/toolchain.cmake.in:115]: Why have these become env vars?

**Comment by hebasto** [comment]
Rebased to resolve conflicts.

File: depends/toolchain.cmake.in (hunk 2 above)
Comment (hebasto) [reply depends/toolchain.cmake.in:115]: They have always been:
- in the master branch: https://github.com/hebasto/bitcoin/blob/67c0d93982ad214f5e0c9509e9dc3d6d792ad97c/configure.ac#L22-L30
- in the current staging branch: https://github.com/hebasto/bitcoin/blob/f9b43f72e0d3723c56ffd83a7da52f6ed1980f67/cmake/module/CrossPkgConfig.cmake#L11-L12
- and in the pkg-config [documentation](https://manpages.debian.org/bullseye/pkg-config/pkg-config.1.en.html):
> `PKG_CONFIG_PATH`
> A colon-separated (on Windows, semicolon-separated) list of directories to search for .pc files. The default directory will always be searched after searching the path; the default is libdir/pkgconfig:datadir/pkgconfig where libdir is the libdir for pkg-config and datadir is the datadir for pkg-config when it was installed.
>
> `PKG_CONFIG_LIBDIR`
> Replaces the default `pkg-config` search directory, usually `/usr/lib/pkgconfig:/usr/share/pkgconfig`.

It's worth mentioning that [`set(ENV{<variable>} [<value>])`](https://cmake.org/cmake/help/latest/command/set.html#env) command:
> affects only the current CMake process, not the process from which CMake was called, nor the system environment at large, nor the environment of subsequent build or test processes.

File: depends/patches/libevent/fix_mingw_link.patch [hunk 3]
```diff
@@ -0,0 +1,24 @@
+commit d108099913c5fdbe518f3f4d711f248f8522bd10
+Author: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>
+Date:   Mon Apr 22 06:39:35 2024 +0100
+
+    build: Add `Iphlpapi` to `Libs.private` in `*.pc` files on Windows
+    
+    It has been required since https://github.com/libevent/libevent/pull/923
+    at least for the `if_nametoindex` call.
+
+    See https://github.com/libevent/libevent/pull/1622.
```
[the remark this answers is not in the document]
Comment (hebasto) [reply depends/patches/libevent/fix_mingw_link.patch:10]: Right.

File: depends/toolchain.cmake.in (hunk 2 above)
Comment (hebasto) [reply depends/toolchain.cmake.in:115]: The `PKG_CONFIG_ARGN` sets a list of arguments to additionally pass to pkg-config. It is used to pass the `--static` option.

The `PKG_CONFIG_LIBDIR` and `PKG_CONFIG_PATH` affect the pkg-config behaviour in a way that is not achievable by other means according to pkg-config docs.

> that's to work around an autotools quirk

I can't see it as a workaround. Everything follows documentation.

Perhaps, I fail to grasp your idea. Maybe it is possible to put it in some (pseudo)code?

**Comment by theuni** [comment]
```
# TODO: Not all targets, which will be added in the future,
#       require libevent. Therefore, a proper check will be
#       appropriate here.
```
Maybe wait until https://github.com/bitcoin/bitcoin/pull/29904 is merged for this PR, then address this comment? At that point, afaik this is only needed for bitcoin_node and bitcoin-cli?

**Review by theuni — APPROVED** [review]
Tested on macOS. I think there's still some work to do here, but this is an improvement over the current behavior.

ACK 20f83ae67c873cdaabe851ee957af4139fc41036

File: depends/patches/libevent/fix_mingw_link.patch (hunk 3 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply depends/patches/libevent/fix_mingw_link.patch:10]: Why? The patch is being added now.

**Pull request merged**